gethostbyaddr does not support IPv6 on Solaris; getipnodebyaddr
should be used instead.

This was offered upstream:
https://github.com/python/cpython/issues/157497
https://github.com/python/cpython/pull/157498

--- Python-3.13.15/Modules/socketmodule.c
+++ Python-3.13.15/Modules/socketmodule.c
@@ -6093,6 +6093,20 @@ socket_gethostbyaddr(PyObject *self, PyO
         PyErr_SetString(PyExc_OSError, "unsupported address family");
         goto finally;
     }
+#if defined(__sun) && defined(ENABLE_IPV6)
+    /* Solaris gethostbyaddr supports AF_INET only; use platform
+       specific and AF_INET6 capable getipnodebyaddr instead. */
+    if (af == AF_INET6) {
+        Py_BEGIN_ALLOW_THREADS
+        h = getipnodebyaddr(ap, al, af, &h_error);
+        Py_END_ALLOW_THREADS
+        ret = gethost_common(state, h, SAS2SA(&addr), sizeof(addr), af, h_error);
+        if (h != NULL) {
+            freehostent(h);
+        }
+        goto finally;
+    }
+#endif
     Py_BEGIN_ALLOW_THREADS
 #ifdef HAVE_GETHOSTBYNAME_R
 #if   defined(HAVE_GETHOSTBYNAME_R_6_ARG)
